// web3x/contract/tx-deploy.js
"use strict";
/*
  This file is part of web3x.

  web3x is free software: you can redistribute it and/or modify
  it under the terms of the GNU Lesser General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  web3x is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public License
  along with web3x.  If not, see <http://www.gnu.org/licenses/>.
*/
Object.defineProperty(exports, "__esModule", { value: true });
const utils_1 = require("../utils");
const sent_deploy_contract_tx_1 = require("./sent-deploy-contract-tx");
class TxDeploy {
    constructor(eth, contractEntry, contractAbi, deployData, args = [], defaultOptions = {}, onDeployed) {
        this.eth = eth;
        this.contractEntry = contractEntry;
        this.contractAbi = contractAbi;
        this.deployData = deployData;
        this.args = args;
        this.defaultOptions = defaultOptions;
        this.onDeployed = onDeployed;
    }
    async estimateGas(options = {}) {
        return await this.eth.estimateGas(this.getTx(options));
    }
    send(options) {
        const tx = this.getTx(options);
        if (!this.contractEntry.payable && tx.value > 0) {
            throw new Error('Can not send value to non-payable constructor.');
        }
        const promise = this.eth.sendTransaction(tx).getTxHash();
        return new sent_deploy_contract_tx_1.SentDeployContractTx(this.eth, this.contractAbi, promise, this.onDeployed);
    }
    getSendRequestPayload(options) {
        return this.eth.request.sendTransaction(this.getTx(options));
    }
    getTx(options) {
        return {
            from: options.from || this.defaultOptions.from,
            gasPrice: options.gasPrice || this.defaultOptions.gasPrice,
            gas: options.gas || this.defaultOptions.gas,
            value: options.value,
            data: this.encodeABI(),
        };
    }
    encodeABI() {
        return utils_1.hexToBuffer(this.deployData + this.contractEntry.encodeParameters(this.args).replace('0x', ''));
    }
}
exports.TxDeploy = TxDeploy;
