// web3x/contract/tx.js
"use strict";
/*
  This file is part of web3x.

  web3x is free software: you can redistribute it and/or modify
  it under the terms of the GNU Lesser General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  web3x is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public License
  along with web3x.  If not, see <http://www.gnu.org/licenses/>.
*/
Object.defineProperty(exports, "__esModule", { value: true });
const sent_contract_tx_1 = require("./sent-contract-tx");
class Tx {
    constructor(eth, contractEntry, contractAbi, contractAddress, args = [], defaultOptions = {}) {
        this.eth = eth;
        this.contractEntry = contractEntry;
        this.contractAbi = contractAbi;
        this.contractAddress = contractAddress;
        this.args = args;
        this.defaultOptions = defaultOptions;
    }
    async estimateGas(options = {}) {
        return await this.eth.estimateGas(this.getTx(options));
    }
    async call(options = {}, block) {
        const result = await this.eth.call(this.getTx(options), block);
        return this.contractEntry.decodeReturnValue(result);
    }
    getCallRequestPayload(options, block) {
        const result = this.eth.request.call(this.getTx(options), block);
        return {
            ...result,
            format: (result) => this.contractEntry.decodeReturnValue(result),
        };
    }
    send(options) {
        const tx = this.getTx(options);
        if (!this.contractEntry.payable && tx.value !== undefined && tx.value > 0) {
            throw new Error('Can not send value to non-payable contract method.');
        }
        const promise = this.eth.sendTransaction(tx).getTxHash();
        return new sent_contract_tx_1.SentContractTx(this.eth, this.contractAbi, promise);
    }
    getSendRequestPayload(options) {
        return this.eth.request.sendTransaction(this.getTx(options));
    }
    encodeABI() {
        return this.contractEntry.encodeABI(this.args);
    }
    getTx(options = {}) {
        return {
            to: this.contractAddress,
            from: options.from || this.defaultOptions.from,
            gasPrice: options.gasPrice || this.defaultOptions.gasPrice,
            gas: options.gas || this.defaultOptions.gas,
            value: options.value,
            data: this.encodeABI(),
        };
    }
}
exports.Tx = Tx;
