// web3x/evm/tx/tx-receipt.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const bigint_buffer_1 = require("bigint-buffer");
const bn_js_1 = tslib_1.__importDefault(require("bn.js"));
const rlp = tslib_1.__importStar(require("rlp"));
const address_1 = require("../../address");
function serializeTxReceipt(txReceipt) {
    const { cumulativeGasUsed, logs, logsBloomFilter, status } = txReceipt;
    return rlp.encode([
        status ? new bn_js_1.default(1) : new bn_js_1.default(0),
        bigint_buffer_1.toBufferBE(cumulativeGasUsed, 32),
        logsBloomFilter,
        logs.map(log => [log.address.toBuffer(), log.topics, log.data]),
    ]);
}
exports.serializeTxReceipt = serializeTxReceipt;
function deserializeTxReceipt(data) {
    const bufs = rlp.decode(data);
    const logs = bufs[3];
    return {
        status: bigint_buffer_1.toBigIntBE(bufs[0]) > 0,
        cumulativeGasUsed: bigint_buffer_1.toBigIntBE(bufs[1]),
        logsBloomFilter: bufs[2],
        logs: logs.map(logBufs => ({
            address: new address_1.Address(logBufs[0]),
            topics: logBufs[1],
            data: logBufs[2],
        })),
    };
}
exports.deserializeTxReceipt = deserializeTxReceipt;
