// web3x/evm/tx/tx.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const bigint_buffer_1 = require("bigint-buffer");
const rlp = tslib_1.__importStar(require("rlp"));
const address_1 = require("../../address");
const account_1 = tslib_1.__importDefault(require("../../eth-lib/account"));
const bytes_1 = tslib_1.__importDefault(require("../../eth-lib/bytes"));
const utils_1 = require("../../utils");
function serializeTx(tx) {
    const { nonce, gasPrice, gasLimit, to, value, dataOrInit, v, r, s } = tx;
    return rlp.encode([
        bigint_buffer_1.toBufferBE(nonce, 32),
        bigint_buffer_1.toBufferBE(gasPrice, 32),
        bigint_buffer_1.toBufferBE(gasLimit, 32),
        to ? to.toBuffer() : Buffer.of(),
        bigint_buffer_1.toBufferBE(value, 32),
        dataOrInit,
        utils_1.hexToBuffer(v),
        utils_1.hexToBuffer(r),
        utils_1.hexToBuffer(s),
    ]);
}
exports.serializeTx = serializeTx;
function deserializeTx(data) {
    const bufs = rlp.decode(data);
    return {
        nonce: bigint_buffer_1.toBigIntBE(bufs[0]),
        gasPrice: bigint_buffer_1.toBigIntBE(bufs[1]),
        gasLimit: bigint_buffer_1.toBigIntBE(bufs[2]),
        to: bufs[3].length ? new address_1.Address(bufs[3]) : undefined,
        value: bigint_buffer_1.toBigIntBE(bufs[4]),
        dataOrInit: bufs[5],
        v: utils_1.bufferToHex(bufs[6]),
        r: utils_1.bufferToHex(bufs[7]),
        s: utils_1.bufferToHex(bufs[8]),
    };
}
exports.deserializeTx = deserializeTx;
const bigIntToHex = n => '0x' + n.toString(16);
function recoverTransaction(tx) {
    const { to, nonce, gasPrice, gasLimit, value, dataOrInit, v, r, s } = tx;
    const signature = bytes_1.default.flatten([bytes_1.default.pad(32, r), bytes_1.default.pad(32, s), v]);
    const recovery = Number(v);
    const extraData = recovery < 35 ? [] : [bytes_1.default.fromNumber((recovery - 35) >> 1), '0x', '0x'];
    const signingData = [
        bytes_1.default.fromNat(bigIntToHex(nonce)),
        bytes_1.default.fromNat(bigIntToHex(gasPrice)),
        bytes_1.default.fromNat(bigIntToHex(gasLimit)),
        to ? to.toString().toLowerCase() : '0x',
        bytes_1.default.fromNat(bigIntToHex(value)),
        bytes_1.default.fromNat(utils_1.bufferToHex(dataOrInit)),
        ...extraData,
    ];
    const signingDataHex = rlp.encode(signingData);
    return address_1.Address.fromString(account_1.default.recover(utils_1.sha3(signingDataHex), signature));
}
exports.recoverTransaction = recoverTransaction;
