// web3x/shh/shh.js
"use strict";
/*
  This file is part of web3x.

  web3x is free software: you can redistribute it and/or modify
  it under the terms of the GNU Lesser General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  web3x is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public License
  along with web3x.  If not, see <http://www.gnu.org/licenses/>.
*/
Object.defineProperty(exports, "__esModule", { value: true });
const subscriptions_1 = require("../subscriptions");
class Shh {
    constructor(provider) {
        this.provider = provider;
        this.request = new ShhRequestPayloads();
    }
    async send({ method, params, format }) {
        return format(await this.provider.send(method, params));
    }
    async getVersion() {
        const payload = this.request.getVersion();
        return payload.format(await this.send(payload));
    }
    async getInfo() {
        const payload = this.request.getInfo();
        return payload.format(await this.send(payload));
    }
    async setMaxMessageSize(size) {
        const payload = this.request.setMaxMessageSize(size);
        return payload.format(await this.send(payload));
    }
    async setMinPoW(pow) {
        const payload = this.request.setMinPoW(pow);
        return payload.format(await this.send(payload));
    }
    async markTrustedPeer(enode) {
        const payload = this.request.markTrustedPeer(enode);
        return payload.format(await this.send(payload));
    }
    async newKeyPair() {
        const payload = this.request.newKeyPair();
        return payload.format(await this.send(payload));
    }
    async addPrivateKey(privateKey) {
        const payload = this.request.addPrivateKey(privateKey);
        return payload.format(await this.send(payload));
    }
    async deleteKeyPair(id) {
        const payload = this.request.deleteKeyPair(id);
        return payload.format(await this.send(payload));
    }
    async hasKeyPair(id) {
        const payload = this.request.hasKeyPair(id);
        return payload.format(await this.send(payload));
    }
    async getPublicKey(id) {
        const payload = this.request.getPublicKey(id);
        return payload.format(await this.send(payload));
    }
    async getPrivateKey(id) {
        const payload = this.request.getPrivateKey(id);
        return payload.format(await this.send(payload));
    }
    async newSymKey() {
        const payload = this.request.newSymKey();
        return payload.format(await this.send(payload));
    }
    async addSymKey(symKey) {
        const payload = this.request.addSymKey(symKey);
        return payload.format(await this.send(payload));
    }
    async generateSymKeyFromPassword(password) {
        const payload = this.request.generateSymKeyFromPassword(password);
        return payload.format(await this.send(payload));
    }
    async hasSymKey(id) {
        const payload = this.request.hasSymKey(id);
        return payload.format(await this.send(payload));
    }
    async getSymKey(id) {
        const payload = this.request.getSymKey(id);
        return payload.format(await this.send(payload));
    }
    async deleteSymKey(id) {
        const payload = this.request.deleteSymKey(id);
        return payload.format(await this.send(payload));
    }
    async newMessageFilter(options) {
        const payload = this.request.newMessageFilter(options);
        return payload.format(await this.send(payload));
    }
    async getFilterMessages(id) {
        const payload = this.request.getFilterMessages(id);
        return payload.format(await this.send(payload));
    }
    async deleteMessageFilter(id) {
        const payload = this.request.deleteMessageFilter(id);
        return payload.format(await this.send(payload));
    }
    async post(post) {
        const payload = this.request.post(post);
        return payload.format(await this.send(payload));
    }
    subscribeMessages(options) {
        return new subscriptions_1.Subscription('shh', 'messages', [options], this.provider, (message, sub) => sub.emit('data', message));
    }
    subscribe(type, options) {
        switch (type) {
            case 'messages':
                return this.subscribeMessages(options);
            default:
                throw new Error(`Unknown subscription type: ${type}`);
        }
    }
}
exports.Shh = Shh;
