// web3x/wallet/wallet.js
"use strict";
/*
  This file is part of web3x.

  web3x is free software: you can redistribute it and/or modify
  it under the terms of the GNU Lesser General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  web3x is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public License
  along with web3x.  If not, see <http://www.gnu.org/licenses/>.
*/
Object.defineProperty(exports, "__esModule", { value: true });
const util_1 = require("util");
const account_1 = require("../account");
const encryption_1 = require("../utils/encryption");
const DEFAULT_KEY_NAME = 'web3js_wallet';
class Wallet {
    constructor(numberOfAccounts = 0) {
        this.length = 0;
        this.accounts = [];
        this.create(numberOfAccounts);
    }
    static fromMnemonic(mnemonic, numberOfAccounts) {
        const wallet = new Wallet();
        for (let i = 0; i < numberOfAccounts; ++i) {
            const path = `m/44'/60'/0'/0/${i}`;
            wallet.add(account_1.Account.createFromMnemonicAndPath(mnemonic, path));
        }
        return wallet;
    }
    static fromSeed(seed, numberOfAccounts) {
        const wallet = new Wallet();
        for (let i = 0; i < numberOfAccounts; ++i) {
            const path = `m/44'/60'/0'/0/${i}`;
            wallet.add(account_1.Account.createFromSeedAndPath(seed, path));
        }
        return wallet;
    }
    static async fromKeystores(encryptedWallet, password) {
        const wallet = new Wallet();
        await wallet.decrypt(encryptedWallet, password);
        return wallet;
    }
    static async fromLocalStorage(password, keyName = DEFAULT_KEY_NAME) {
        if (!localStorage) {
            return;
        }
        const keystoreStr = localStorage.getItem(keyName);
        if (!keystoreStr) {
            return;
        }
        try {
            return Wallet.fromKeystores(JSON.parse(keystoreStr), password);
        }
        catch (e) {
            return;
        }
    }
    create(numberOfAccounts, entropy) {
        for (let i = 0; i < numberOfAccounts; ++i) {
            this.add(account_1.Account.create(entropy).privateKey);
        }
        return this.accounts;
    }
    get(addressOrIndex) {
        if (util_1.isNumber(addressOrIndex)) {
            return this.accounts[addressOrIndex];
        }
        return this.accounts.find(a => a && a.address.toString().toLowerCase() === addressOrIndex.toString().toLowerCase());
    }
    indexOf(addressOrIndex) {
        if (util_1.isNumber(addressOrIndex)) {
            return addressOrIndex;
        }
        return this.accounts.findIndex(a => a.address.toString().toLowerCase() === addressOrIndex.toString().toLowerCase());
    }
    add(accountOrKey) {
        const account = Buffer.isBuffer(accountOrKey) ? account_1.Account.fromPrivate(accountOrKey) : accountOrKey;
        const existing = this.get(account.address);
        if (existing) {
            return existing;
        }
        const index = this.findSafeIndex();
        this.accounts[index] = account;
        this.length++;
        return account;
    }
    remove(addressOrIndex) {
        const index = this.indexOf(addressOrIndex);
        if (index === -1) {
            return false;
        }
        delete this.accounts[index];
        this.length--;
        return true;
    }
    clear() {
        this.accounts = [];
        this.length = 0;
    }
    encrypt(password, options) {
        return Promise.all(this.currentIndexes().map(index => this.accounts[index].encrypt(password, options)));
    }
    async decrypt(encryptedWallet, password) {
        const decrypted = await Promise.all(encryptedWallet.map(keystore => encryption_1.decrypt(keystore, password)));
        decrypted.forEach(account => {
            if (!account) {
                throw new Error("Couldn't decrypt accounts. Password wrong?");
            }
            this.add(account);
        });
        return this.accounts;
    }
    async saveToLocalStorage(password, keyName = DEFAULT_KEY_NAME) {
        if (!localStorage) {
            return false;
        }
        localStorage.setItem(keyName, JSON.stringify(await this.encrypt(password)));
        return true;
    }
    findSafeIndex(pointer = 0) {
        while (this.accounts[pointer]) {
            ++pointer;
        }
        return pointer;
    }
    currentIndexes() {
        return Object.keys(this.accounts).map(key => +key);
    }
    currentAddresses() {
        return Object.entries(this.accounts).map(([, account]) => account.address);
    }
}
exports.Wallet = Wallet;
