// web3x/ethers/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const bignumber_1 = require("./bignumber");
const AddressZero = '0x0000000000000000000000000000000000000000';
exports.AddressZero = AddressZero;
const HashZero = '0x0000000000000000000000000000000000000000000000000000000000000000';
exports.HashZero = HashZero;
// NFKD (decomposed)
//const EtherSymbol = '\uD835\uDF63';
// NFKC (composed)
const EtherSymbol = '\u039e';
exports.EtherSymbol = EtherSymbol;
const NegativeOne = bignumber_1.bigNumberify(-1);
exports.NegativeOne = NegativeOne;
const Zero = bignumber_1.bigNumberify(0);
exports.Zero = Zero;
const One = bignumber_1.bigNumberify(1);
exports.One = One;
const Two = bignumber_1.bigNumberify(2);
exports.Two = Two;
const WeiPerEther = bignumber_1.bigNumberify('1000000000000000000');
exports.WeiPerEther = WeiPerEther;
const MaxUint256 = bignumber_1.bigNumberify('0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff');
exports.MaxUint256 = MaxUint256;
