// web3x/contract/abi/contract-function-entry.js
"use strict";
/*
  This file is part of web3x.

  web3x is free software: you can redistribute it and/or modify
  it under the terms of the GNU Lesser General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  web3x is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public License
  along with web3x.  If not, see <http://www.gnu.org/licenses/>.
*/
Object.defineProperty(exports, "__esModule", { value: true });
const utils_1 = require("../../utils");
const abi_coder_1 = require("../abi-coder");
const contract_entry_1 = require("./contract-entry");
class ContractFunctionEntry extends contract_entry_1.ContractEntry {
    constructor(entry) {
        super(entry);
        this.signature = abi_coder_1.abiCoder.encodeFunctionSignature(abi_coder_1.abiCoder.abiMethodToString(entry));
    }
    get constant() {
        return this.entry.stateMutability === 'view' || this.entry.stateMutability === 'pure' || this.entry.constant;
    }
    get payable() {
        return this.entry.stateMutability === 'payable' || this.entry.payable;
    }
    numArgs() {
        return this.entry.inputs ? this.entry.inputs.length : 0;
    }
    decodeReturnValue(returnValue) {
        if (!returnValue) {
            return null;
        }
        const result = abi_coder_1.abiCoder.decodeParameters(this.entry.outputs, returnValue);
        if (result.__length__ === 1) {
            return result[0];
        }
        else {
            delete result.__length__;
            return result;
        }
    }
    encodeABI(args) {
        return utils_1.hexToBuffer(this.signature + this.encodeParameters(args).replace('0x', ''));
    }
    encodeParameters(args) {
        return abi_coder_1.abiCoder.encodeParameters(this.entry.inputs || [], args);
    }
}
exports.ContractFunctionEntry = ContractFunctionEntry;
