// web3x/evm/world/account-state.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const bigint_buffer_1 = require("bigint-buffer");
const bn_js_1 = tslib_1.__importDefault(require("bn.js"));
const rlp = tslib_1.__importStar(require("rlp"));
class AccountState {
    constructor(nonce = BigInt(0), balance = BigInt(0), storageRoot = Buffer.of(), codeHash = Buffer.of()) {
        this.nonce = nonce;
        this.balance = balance;
        this.storageRoot = storageRoot;
        this.codeHash = codeHash;
    }
    static fromRlp(data) {
        const account = rlp.decode(data);
        return new AccountState(account[0].length ? bigint_buffer_1.toBigIntBE(account[0]) : BigInt(0), bigint_buffer_1.toBigIntBE(account[1]), account[2], account[3]);
    }
    toRlp() {
        return rlp.encode([
            new bn_js_1.default(this.nonce.toString()),
            new bn_js_1.default(this.balance.toString()),
            this.storageRoot,
            this.codeHash,
        ]);
    }
}
exports.AccountState = AccountState;
