// web3x/evm/world/evm-account.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const bigint_buffer_1 = require("bigint-buffer");
const rlp = tslib_1.__importStar(require("rlp"));
const address_1 = require("../../address");
const utils_1 = require("../../utils");
const trie_1 = require("../trie");
const vm_1 = require("../vm");
function EvmAccountFactory(address, nonce, balance, storage, code) {
    switch (address.toString()) {
        case '0x0000000000000000000000000000000000000001':
            return new EvmEcdsaRecoveryAccount(address, nonce, balance, storage, code);
        default:
            return new EvmAccount(address, nonce, balance, storage, code);
    }
}
exports.EvmAccountFactory = EvmAccountFactory;
class EvmAccount {
    constructor(address, nonce, balance, storage, code) {
        this.address = address;
        this.nonce = nonce;
        this.balance = balance;
        this.storage = storage;
        this.code = code;
    }
    nextContractAddress() {
        return new address_1.Address(utils_1.sha3Buffer(rlp.encode([this.address.toBuffer(), bigint_buffer_1.toBufferBE(this.nonce - BigInt(1), 32)])).slice(12));
    }
    async run(context) {
        if (context.code.length === 0) {
            return context;
        }
        return await vm_1.run(context);
    }
    isEmpty() {
        return this.code.length === 0 && this.nonce === BigInt(0) && this.balance === BigInt(0);
    }
}
exports.EvmAccount = EvmAccount;
class EvmEcdsaRecoveryAccount extends EvmAccount {
    constructor(address, nonce, balance, storage, code) {
        super(address, nonce, balance, storage, code);
    }
    static fromDb(db) {
        return new EvmEcdsaRecoveryAccount(address_1.Address.fromString('0x0000000000000000000000000000000000000001'), BigInt(0), BigInt(0), new trie_1.Trie(db), Buffer.of());
    }
    async run(callContext) {
        const { calldata } = callContext;
        const h = calldata.slice(0, 32);
        const v = calldata.slice(32, 64);
        const r = calldata.slice(64, 96);
        const s = calldata.slice(96, 128);
        const result = utils_1.recover(utils_1.bufferToHex(h), utils_1.bufferToHex(v), utils_1.bufferToHex(r), utils_1.bufferToHex(s), true).toBuffer32();
        callContext.halt = true;
        callContext.returned = result;
        return callContext;
    }
}
exports.EvmEcdsaRecoveryAccount = EvmEcdsaRecoveryAccount;
