// web3x/evm/opcodes/call.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const bigint_buffer_1 = require("bigint-buffer");
const address_1 = require("../../address");
const message_call_1 = require("../vm/message-call");
class CallOp {
    constructor() {
        this.code = 0xf1;
        this.mnemonic = 'CALL';
        this.description = 'Message-call into an account';
        this.gas = 0;
        this.bytes = 1;
    }
    toString(params) {
        return `${this.mnemonic}`;
    }
    async handle(context) {
        const { stack, worldState, memory, origin, executor, callDepth, modify } = context;
        const gas = stack.pop();
        const addr = stack.pop();
        const value = stack.pop();
        const inOffset = stack.pop();
        const inSize = stack.pop();
        const retOffset = stack.pop();
        const retSize = stack.pop();
        const recipient = new address_1.Address(bigint_buffer_1.toBufferBE(addr, 20));
        const calldata = memory.loadN(inOffset, Number(inSize));
        const { txSubstrate, status, returned } = await message_call_1.messageCall(worldState, executor, origin, recipient, recipient, value, gas, calldata, callDepth + 1, modify);
        if (!status) {
            context.stack.push(BigInt(0));
        }
        else {
            context.stack.push(BigInt(1));
            context.txSubstrate.logs.push(...txSubstrate.logs);
            context.memory.storeN(retOffset, returned.slice(0, Number(retSize)));
            context.lastReturned = returned;
        }
        context.ip += this.bytes;
    }
}
exports.Call = new CallOp();
