// web3x/evm/opcodes/log.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const bigint_buffer_1 = require("bigint-buffer");
class LogOp {
    constructor(code, topics) {
        this.code = code;
        this.topics = topics;
        this.bytes = 1;
        this.gas = 375 * (topics + 1);
        this.mnemonic = 'LOG' + topics;
        this.description = `Append log record with ${topics} topics`;
    }
    toString(params) {
        return `${this.mnemonic}`;
    }
    handle(context) {
        const offset = context.stack.pop();
        const length = context.stack.pop();
        const data = context.memory.loadN(offset, Number(length));
        const topics = [];
        for (let i = 0; i < this.topics; i++) {
            topics.push(bigint_buffer_1.toBufferBE(context.stack.pop(), 32));
        }
        context.txSubstrate.logs.push({ data, topics, address: context.executor });
        context.ip += this.bytes;
    }
}
exports.LogOps = new Array(5).fill(0).map((_, i) => new LogOp(0xa0 + i, i));
