// @bloomprotocol/attestations-lib/dist/index.js
"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
var AttestationTypes_1 = require("./src/AttestationTypes");
exports.AttestationTypeID = AttestationTypes_1.AttestationTypeID;
exports.AttestationTypes = AttestationTypes_1.AttestationTypes;
exports.AttestationTypesByID = AttestationTypes_1.AttestationTypesByID;
exports.AttestationTypeNames = AttestationTypes_1.AttestationTypeNames;
exports.AttestationTypeIDs = AttestationTypes_1.AttestationTypeIDs;
exports.getAttestationTypeAttrib = AttestationTypes_1.getAttestationTypeAttrib;
exports.getAttestationTypeStr = AttestationTypes_1.getAttestationTypeStr;
exports.getBloomIDStrength = AttestationTypes_1.getBloomIDStrength;
exports.getFormattedName = AttestationTypes_1.getFormattedName;
exports.AttestationStatus = AttestationTypes_1.AttestationStatus;
var HashingLogic = __importStar(require("./src/HashingLogic"));
exports.HashingLogic = HashingLogic;
