// web3x/evm/vm/evm-context.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const address_1 = require("../../address");
const memory_1 = require("../memory");
const stack_1 = require("../stack");
const trie_1 = require("../trie");
const tx_substrate_1 = require("../tx/tx-substrate");
class EvmContext {
    constructor(worldState, code = Buffer.of(), calldata = Buffer.of(), origin = address_1.Address.ZERO, caller = address_1.Address.ZERO, executor = address_1.Address.ZERO, value = BigInt(0), gas = BigInt(0), storage = new trie_1.Trie(), callDepth = 0, modify = true, txSubstrate = new tx_substrate_1.TxSubstrate()) {
        this.worldState = worldState;
        this.code = code;
        this.calldata = calldata;
        this.origin = origin;
        this.caller = caller;
        this.executor = executor;
        this.value = value;
        this.gas = gas;
        this.storage = storage;
        this.callDepth = callDepth;
        this.modify = modify;
        this.txSubstrate = txSubstrate;
        this.stack = new stack_1.Stack();
        this.memory = new memory_1.EvmMemory();
        this.ip = 0;
        this.halt = false;
        this.reverted = false;
        this.revertInstruction = 0;
        this.returned = Buffer.of();
        this.lastReturned = Buffer.of();
    }
}
exports.EvmContext = EvmContext;
