// web3x/evm/vm/gas.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var Gas;
(function (Gas) {
    Gas[Gas["GZERO"] = 0] = "GZERO";
    Gas[Gas["GBASE"] = 2] = "GBASE";
    Gas[Gas["GVERYLOW"] = 3] = "GVERYLOW";
    Gas[Gas["GLOW"] = 5] = "GLOW";
    Gas[Gas["GMID"] = 8] = "GMID";
    Gas[Gas["GHIGH"] = 10] = "GHIGH";
    Gas[Gas["GEXTCODE"] = 700] = "GEXTCODE";
    Gas[Gas["GBALANCE"] = 400] = "GBALANCE";
    Gas[Gas["GSLOAD"] = 200] = "GSLOAD";
    Gas[Gas["GJUMPDEST"] = 1] = "GJUMPDEST";
    Gas[Gas["GSSET"] = 20000] = "GSSET";
    Gas[Gas["GSRESET"] = 5000] = "GSRESET";
    Gas[Gas["RSCLEAR"] = 15000] = "RSCLEAR";
    Gas[Gas["RSELFDESTRUCT"] = 24000] = "RSELFDESTRUCT";
    Gas[Gas["GSELFDESTRUCT"] = 5000] = "GSELFDESTRUCT";
    Gas[Gas["GCREATE"] = 32000] = "GCREATE";
    Gas[Gas["GCODEDEPOSIT"] = 200] = "GCODEDEPOSIT";
    Gas[Gas["GCALL"] = 700] = "GCALL";
    Gas[Gas["GCALLVALUE"] = 9000] = "GCALLVALUE";
    Gas[Gas["GCALLSTIPEND"] = 2300] = "GCALLSTIPEND";
    Gas[Gas["GNEWACCOUNT"] = 25000] = "GNEWACCOUNT";
    Gas[Gas["GEXP"] = 10] = "GEXP";
    Gas[Gas["GEXPBYTE"] = 50] = "GEXPBYTE";
    Gas[Gas["GMEMORY"] = 3] = "GMEMORY";
    Gas[Gas["GTXCREATE"] = 32000] = "GTXCREATE";
    Gas[Gas["GTXDATAZERO"] = 4] = "GTXDATAZERO";
    Gas[Gas["GTXDATANONZERO"] = 68] = "GTXDATANONZERO";
    Gas[Gas["GTRANSACTION"] = 21000] = "GTRANSACTION";
    Gas[Gas["GLOG"] = 375] = "GLOG";
    Gas[Gas["GLOGDATA"] = 8] = "GLOGDATA";
    Gas[Gas["GLOGTOPIC"] = 375] = "GLOGTOPIC";
    Gas[Gas["GSHA3"] = 30] = "GSHA3";
    Gas[Gas["GSHA3WORD"] = 6] = "GSHA3WORD";
    Gas[Gas["GCOPY"] = 3] = "GCOPY";
    Gas[Gas["GBLOCKHASH"] = 20] = "GBLOCKHASH";
    Gas[Gas["GQUADDIVISOR"] = 100] = "GQUADDIVISOR";
})(Gas = exports.Gas || (exports.Gas = {}));
