// web3x/contract/abi-coder/index.js
"use strict";
/*
  This file is part of web3x.

  web3x is free software: you can redistribute it and/or modify
  it under the terms of the GNU Lesser General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  web3x is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public License
  along with web3x.  If not, see <http://www.gnu.org/licenses/>.
*/
Object.defineProperty(exports, "__esModule", { value: true });
const util_1 = require("util");
const abi_coder_1 = require("../../ethers/abi-coder");
const utils_1 = require("../../utils");
/**
 * ABICoder prototype should be used to encode/decode solidity params of any type
 */
class ABICoder {
    constructor() {
        this.ethersAbiCoder = new abi_coder_1.AbiCoder((type, value) => {
            if (type.match(/^u?int/) && !util_1.isArray(value) && (!util_1.isObject(value) || value.constructor.name !== 'BN')) {
                return value.toString();
            }
            return value;
        });
    }
    /**
     * Encodes the function name to its ABI representation, which are the first 4 bytes of the sha3 of the function name including  types.
     *
     * @method encodeFunctionSignature
     * @param {String|Object} functionName
     * @return {String} encoded function name
     */
    encodeFunctionSignature(functionName) {
        if (util_1.isObject(functionName)) {
            functionName = this.abiMethodToString(functionName);
        }
        return utils_1.sha3(functionName).slice(0, 10);
    }
    /**
     * Encodes the function name to its ABI representation, which are the first 4 bytes of the sha3 of the function name including  types.
     *
     * @method encodeEventSignature
     * @param {String|Object} functionName
     * @return {String} encoded function name
     */
    encodeEventSignature(functionName) {
        if (util_1.isObject(functionName)) {
            functionName = this.abiMethodToString(functionName);
        }
        return utils_1.sha3(functionName);
    }
    /**
     * Should be used to encode plain param
     *
     * @method encodeParameter
     * @param {String} type
     * @param {Object} param
     * @return {String} encoded plain param
     */
    encodeParameter(type, param) {
        return this.encodeParameters([type], [param]);
    }
    /**
     * Should be used to encode list of params
     *
     * @method encodeParameters
     * @param {Array} types
     * @param {Array} params
     * @return {String} encoded list of params
     */
    encodeParameters(types, params) {
        return this.ethersAbiCoder.encode(this.mapTypes(types), params);
    }
    /**
     * Encodes a function call from its json interface and parameters.
     *
     * @method encodeFunctionCall
     * @param {Array} jsonInterface
     * @param {Array} params
     * @return {String} The encoded ABI for this function call
     */
    encodeFunctionCall(jsonInterface, params) {
        return (this.encodeFunctionSignature(jsonInterface) +
            this.encodeParameters(jsonInterface.inputs, params).replace('0x', ''));
    }
    /**
     * Should be used to decode bytes to plain param
     *
     * @method decodeParameter
     * @param {String} type
     * @param {String} bytes
     * @return {Object} plain param
     */
    decodeParameter(type, bytes) {
        return this.decodeParameters([type], bytes)[0];
    }
    /**
     * Should be used to decode list of params
     *
     * @method decodeParameter
     * @param {Array} outputs
     * @param {String} bytes
     * @return {Array} array of plain params
     */
    decodeParameters(outputs, bytes) {
        if (!bytes || bytes === '0x' || bytes === '0X') {
            throw new Error("Returned values aren't valid, did it run Out of Gas?");
        }
        const res = this.ethersAbiCoder.decode(this.mapTypes(outputs), '0x' + bytes.replace(/0x/i, ''));
        const returnValue = {};
        returnValue.__length__ = 0;
        outputs.forEach((output, i) => {
            let decodedValue = res[returnValue.__length__];
            decodedValue = decodedValue === '0x' ? null : decodedValue;
            returnValue[i] = decodedValue;
            if (util_1.isObject(output) && output.name) {
                returnValue[output.name] = decodedValue;
            }
            returnValue.__length__++;
        });
        return returnValue;
    }
    /**
     * Decodes events non- and indexed parameters.
     *
     * @method decodeLog
     * @param {Object} inputs
     * @param {String} data
     * @param {Array} topics
     * @return {Array} array of plain params
     */
    decodeLog(inputs, data, topics) {
        topics = util_1.isArray(topics) ? topics : [topics];
        data = data || '';
        const notIndexedInputs = [];
        const indexedParams = [];
        let topicCount = 0;
        // TODO check for anonymous logs?
        inputs.forEach((input, i) => {
            if (input.indexed) {
                indexedParams[i] = ['bool', 'int', 'uint', 'address', 'fixed', 'ufixed'].find(staticType => input.type.indexOf(staticType) !== -1)
                    ? this.decodeParameter(input.type, topics[topicCount])
                    : topics[topicCount];
                topicCount++;
            }
            else {
                notIndexedInputs[i] = input;
            }
        });
        const nonIndexedData = data;
        const notIndexedParams = nonIndexedData && nonIndexedData !== '0x' ? this.decodeParameters(notIndexedInputs, nonIndexedData) : [];
        const returnValue = {};
        returnValue.__length__ = 0;
        inputs.forEach((res, i) => {
            returnValue[i] = res.type === 'string' ? '' : null;
            if (typeof notIndexedParams[i] !== 'undefined') {
                returnValue[i] = notIndexedParams[i];
            }
            if (typeof indexedParams[i] !== 'undefined') {
                returnValue[i] = indexedParams[i];
            }
            if (res.name) {
                returnValue[res.name] = returnValue[i];
            }
            returnValue.__length__++;
        });
        return returnValue;
    }
    /**
     * Map types if simplified format is used
     *
     * @method mapTypes
     * @param {Array} types
     * @return {Array}
     */
    mapTypes(types) {
        const mappedTypes = [];
        types.forEach(type => {
            if (this.isSimplifiedStructFormat(type)) {
                const structName = Object.keys(type)[0];
                mappedTypes.push(Object.assign(this.mapStructNameAndType(structName), {
                    components: this.mapStructToCoderFormat(type[structName]),
                }));
                return;
            }
            mappedTypes.push(type);
        });
        return mappedTypes;
    }
    /**
     * Check if type is simplified struct format
     *
     * @method isSimplifiedStructFormat
     * @param {string | Object} type
     * @returns {boolean}
     */
    isSimplifiedStructFormat(type) {
        return typeof type === 'object' && typeof type.components === 'undefined' && typeof type.name === 'undefined';
    }
    /**
     * Maps the correct tuple type and name when the simplified format in encode/decodeParameter is used
     *
     * @method mapStructNameAndType
     * @param {string} structName
     * @return {{type: string, name: *}}
     */
    mapStructNameAndType(structName) {
        let type = 'tuple';
        if (structName.indexOf('[]') > -1) {
            type = 'tuple[]';
            structName = structName.slice(0, -2);
        }
        return { type, name: structName };
    }
    /**
     * Maps the simplified format in to the expected format of the ABICoder
     *
     * @method mapStructToCoderFormat
     * @param {Object} struct
     * @return {Array}
     */
    mapStructToCoderFormat(struct) {
        const components = [];
        Object.keys(struct).forEach(key => {
            if (typeof struct[key] === 'object') {
                components.push(Object.assign(this.mapStructNameAndType(key), {
                    components: this.mapStructToCoderFormat(struct[key]),
                }));
                return;
            }
            components.push({
                name: key,
                type: struct[key],
            });
        });
        return components;
    }
    /**
     * Should be used to create full function/event name from json abi
     *
     * @method jsonInterfaceMethodToString
     * @param {Object} json
     * @return {String} full function/event name
     */
    abiMethodToString(json) {
        if (util_1.isObject(json) && json.name && json.name.indexOf('(') !== -1) {
            return json.name;
        }
        return json.name + '(' + flattenTypes(false, json.inputs).join(',') + ')';
    }
}
exports.ABICoder = ABICoder;
/**
 * Should be used to flatten json abi inputs/outputs into an array of type-representing-strings
 *
 * @method flattenTypes
 * @param {bool} includeTuple
 * @param {Object} puts
 * @return {Array} parameters as strings
 */
function flattenTypes(includeTuple, puts) {
    // console.log("entered _flattenTypes. inputs/outputs: " + puts)
    const types = [];
    puts.forEach(param => {
        if (typeof param.components === 'object') {
            if (param.type.substring(0, 5) !== 'tuple') {
                throw new Error('components found but type is not tuple; report on GitHub');
            }
            let suffix = '';
            const arrayBracket = param.type.indexOf('[');
            if (arrayBracket >= 0) {
                suffix = param.type.substring(arrayBracket);
            }
            const result = flattenTypes(includeTuple, param.components);
            // console.log("result should have things: " + result)
            if (util_1.isArray(result) && includeTuple) {
                // console.log("include tuple word, and its an array. joining...: " + result.types)
                types.push('tuple(' + result.join(',') + ')' + suffix);
            }
            else if (!includeTuple) {
                // console.log("don't include tuple, but its an array. joining...: " + result)
                types.push('(' + result.join(',') + ')' + suffix);
            }
            else {
                // console.log("its a single type within a tuple: " + result.types)
                types.push('(' + result + ')');
            }
        }
        else {
            // console.log("its a type and not directly in a tuple: " + param.type)
            types.push(param.type);
        }
    });
    return types;
}
exports.abiCoder = new ABICoder();
