// web3x/evm/blockchain/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const bn_js_1 = tslib_1.__importDefault(require("bn.js"));
const events_1 = require("events");
const rlp = tslib_1.__importStar(require("rlp"));
const address_1 = require("../../address");
const utils_1 = require("../../utils");
const trie_1 = require("../trie");
const tx_1 = require("../tx/tx");
const tx_receipt_1 = require("../tx/tx-receipt");
function serializeBlockHeader(blockHeader) {
    return rlp.encode([
        blockHeader.parentHash,
        blockHeader.stateRoot,
        blockHeader.transactionsRoot,
        blockHeader.receiptsRoot,
        blockHeader.number,
    ]);
}
exports.serializeBlockHeader = serializeBlockHeader;
function deserializeBlockHeader(data) {
    const bufs = rlp.decode(data);
    return {
        parentHash: bufs[0],
        sha3Uncles: Buffer.of(),
        miner: address_1.Address.ZERO,
        stateRoot: bufs[1],
        transactionsRoot: bufs[2],
        receiptsRoot: bufs[3],
        logsBloom: Buffer.of(),
        difficulty: BigInt(0),
        number: new bn_js_1.default(bufs[4]).toNumber(),
        gasLimit: BigInt(0),
        gasUsed: BigInt(0),
        timestamp: new Date().getTime(),
        extraData: Buffer.of(),
        mixHash: Buffer.of(),
        nonce: 0,
    };
}
exports.deserializeBlockHeader = deserializeBlockHeader;
class Blockchain extends events_1.EventEmitter {
    constructor(db, blocks) {
        super();
        this.db = db;
        this.blocks = blocks;
    }
    static async fromDb(db) {
        const getChainTip = async () => {
            try {
                return await db.get(Buffer.from('chainTip'));
            }
            catch (err) {
                return null;
            }
        };
        let blockHash = await getChainTip();
        const blocks = [];
        if (blockHash) {
            while (true) {
                const block = deserializeBlockHeader(await db.get(blockHash));
                blocks.unshift(block);
                if (block.parentHash.length === 0) {
                    break;
                }
                blockHash = block.parentHash;
            }
        }
        return new Blockchain(db, blocks);
    }
    async mineTransactions(stateRoot, txs, txReceipts) {
        const parentBlock = this.blocks.length > 0 ? this.blocks[this.blocks.length - 1] : undefined;
        const blockNumber = parentBlock ? parentBlock.number + 1 : 0;
        const serializedTxs = txs.map(tx_1.serializeTx);
        const txHashes = serializedTxs.map(utils_1.sha3Buffer);
        const receiptTrie = new trie_1.Trie(this.db);
        txReceipts.forEach((receipt, i) => receiptTrie.put(utils_1.sha3Buffer(i.toString()), tx_receipt_1.serializeTxReceipt(receipt)));
        const txTrie = new trie_1.Trie(this.db);
        serializedTxs.forEach((tx, i) => txTrie.put(utils_1.sha3Buffer(i.toString()), tx));
        const block = {
            parentHash: parentBlock ? utils_1.sha3Buffer(serializeBlockHeader(parentBlock)) : Buffer.of(),
            sha3Uncles: Buffer.of(),
            miner: address_1.Address.ZERO,
            stateRoot,
            transactionsRoot: txTrie.root,
            receiptsRoot: receiptTrie.root,
            logsBloom: Buffer.of(),
            difficulty: BigInt(0),
            number: blockNumber,
            gasLimit: BigInt(0),
            gasUsed: BigInt(0),
            timestamp: new Date().getTime(),
            extraData: Buffer.of(),
            mixHash: Buffer.of(),
            nonce: 0,
        };
        this.blocks.push(block);
        // Add block to db.
        const serializedBlock = serializeBlockHeader(block);
        const blockHash = utils_1.sha3Buffer(serializedBlock);
        await this.db.put(blockHash, serializedBlock);
        // Record new chaintip.
        await this.db.put(Buffer.from('chainTip'), blockHash);
        // Add lookup for transactions. txHash => [blockHash, txIndex].
        await Promise.all(txHashes.map((txHash, i) => this.db.put(txHash, rlp.encode([blockHash, utils_1.sha3Buffer(i.toString())]))));
        this.emit('newHeads', block, blockHash);
        txReceipts.forEach((receipt, transactionIndex) => {
            const transactionHash = txHashes[transactionIndex];
            const result = {
                blockHash: utils_1.bufferToHex(blockHash),
                blockNumber,
                transactionIndex,
                transactionHash: utils_1.bufferToHex(transactionHash),
            };
            receipt.logs.forEach(({ data, address, topics }, logIndex) => {
                this.emit('logs', { ...result, logIndex, data: utils_1.bufferToHex(data), address, topics: topics.map(utils_1.bufferToHex) });
            });
        });
        return txHashes;
    }
    async getMinedTransaction(txHash) {
        const lookup = rlp.decode(await this.db.get(txHash));
        const blockHeader = deserializeBlockHeader(await this.db.get(lookup[0]));
        const txTrie = new trie_1.Trie(this.db, blockHeader.transactionsRoot);
        return tx_1.deserializeTx(await txTrie.get(lookup[1]));
    }
    async getTransactionReceipt(txHash) {
        try {
            const lookup = rlp.decode(await this.db.get(txHash));
            const blockHeader = deserializeBlockHeader(await this.db.get(lookup[0]));
            const receiptTrie = new trie_1.Trie(this.db, blockHeader.receiptsRoot);
            return tx_receipt_1.deserializeTxReceipt(await receiptTrie.get(lookup[1]));
        }
        catch (err) {
            return null;
        }
    }
    async indexKeyedTrieToArray(trie) {
        const result = [];
        for (let i = 0; true; i++) {
            const data = await trie.get(utils_1.sha3Buffer(i.toString()));
            if (!data) {
                return result;
            }
            result.push(data);
        }
    }
    async getLogs(addresses, topics, fromBlock, toBlock) {
        if (fromBlock < 0 ||
            fromBlock > this.blocks.length - 1 ||
            toBlock < 0 ||
            toBlock > this.blocks.length - 1 ||
            toBlock < fromBlock) {
            throw new Error('Bad fromBlock toBlock range.');
        }
        const blocks = this.blocks.slice(fromBlock, toBlock + 1);
        // TODO: Filter for blocks with appropriate logsBloom.
        const logs = [];
        for (const block of blocks) {
            const blockHash = utils_1.sha3Buffer(serializeBlockHeader(block));
            const receiptTrie = new trie_1.Trie(this.db, block.receiptsRoot);
            const receipts = (await this.indexKeyedTrieToArray(receiptTrie)).map(tx_receipt_1.deserializeTxReceipt);
            for (const [transactionIndex, receipt] of receipts.entries()) {
                /* TODO: Skip receipts that don't match filter.
                if (receipt.logsBloomFilter != topics) {
                  continue;
                }
                */
                const txTrie = new trie_1.Trie(this.db, block.transactionsRoot);
                const txData = await txTrie.get(utils_1.sha3Buffer(transactionIndex.toString()));
                const transactionHash = utils_1.sha3Buffer(txData);
                // Add another filter before map to only push topics of interest.
                logs.push(...receipt.logs
                    .filter(log => addresses.some(a => a.equals(log.address)))
                    .filter(log => topicsMatch(log.topics, topics))
                    .map(log => ({ block, blockHash, log, transactionIndex, transactionHash })));
            }
        }
        return logs;
    }
    getChaintip() {
        return this.blocks[this.blocks.length - 1];
    }
}
exports.Blockchain = Blockchain;
function topicsMatch(logTopics, topics) {
    return logTopics.every((logTopic, index) => {
        const topic = topics[index];
        return topic === null ? true : topic.some(topicMatcher => topicMatcher.equals(logTopic));
    });
}
