// web3x/subscriptions/subscription.js
"use strict";
/*
  This file is part of web3x.

  web3x is free software: you can redistribute it and/or modify
  it under the terms of the GNU Lesser General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  web3x is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public License
  along with web3x.  If not, see <http://www.gnu.org/licenses/>.
*/
Object.defineProperty(exports, "__esModule", { value: true });
const events_1 = require("events");
const util_1 = require("util");
class Subscription extends events_1.EventEmitter {
    constructor(type, subscription, params, provider, callback, subscribeImmediately = true) {
        super();
        this.type = type;
        this.subscription = subscription;
        this.params = params;
        this.provider = provider;
        this.callback = callback;
        if (subscribeImmediately) {
            this.subscribe();
        }
    }
    async subscribe() {
        if (this.id) {
            this.unsubscribe();
        }
        try {
            this.listener = params => this.notificationHandler(params);
            this.provider.on('notification', this.listener);
            this.id = await this.provider.send(`${this.type}_subscribe`, [this.subscription, ...this.params]);
            if (!this.id) {
                throw new Error(`Failed to subscribe to ${this.subscription}.`);
            }
        }
        catch (err) {
            this.emit('error', err, this);
        }
        return this;
    }
    notificationHandler(params) {
        const { subscription, result } = params;
        if (subscription !== this.id) {
            return;
        }
        if (result instanceof Error) {
            this.unsubscribe();
            this.emit('error', result, this);
            return;
        }
        const resultArr = util_1.isArray(result) ? result : [result];
        resultArr.forEach(resultItem => {
            this.callback(resultItem, this);
        });
    }
    unsubscribe() {
        if (this.listener) {
            this.provider.removeListener('notification', this.listener);
        }
        if (this.id) {
            this.provider.send(`${this.type}_unsubscribe`, [this.id]);
        }
        this.id = undefined;
        this.listener = undefined;
    }
}
exports.Subscription = Subscription;
