// @bloomprotocol/attestations-lib/dist/src/AttestationTypeID.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * Should correlate to AttestationLogic permittedTypesList ordering.
 *
 * Example call to verify would look like:
 * var attestationLogic = new AttestationLogic(CONTRACT_ADDRESS);
 * attesationLogic.permittedTypesList.call(0) // should print out 'phone'
 * attesationLogic.permittedTypesList.call(5) // should print out 'id-document'
 */
var AttestationTypeID;
(function (AttestationTypeID) {
    AttestationTypeID[AttestationTypeID["phone"] = 0] = "phone";
    AttestationTypeID[AttestationTypeID["email"] = 1] = "email";
    AttestationTypeID[AttestationTypeID["facebook"] = 2] = "facebook";
    AttestationTypeID[AttestationTypeID["sanctionScreen"] = 3] = "sanctionScreen";
    AttestationTypeID[AttestationTypeID["pepScreen"] = 4] = "pepScreen";
    AttestationTypeID[AttestationTypeID["idDocument"] = 5] = "idDocument";
    AttestationTypeID[AttestationTypeID["google"] = 6] = "google";
    AttestationTypeID[AttestationTypeID["linkedin"] = 7] = "linkedin";
    AttestationTypeID[AttestationTypeID["twitter"] = 8] = "twitter";
})(AttestationTypeID = exports.AttestationTypeID || (exports.AttestationTypeID = {}));
exports.AttestationTypeNames = Object.keys(AttestationTypeID).filter(function (k) { return typeof AttestationTypeID[k] === 'number'; });
var toDashedCase = function (str) {
    return str.replace(/([A-Z])/g, function ($1) {
        return '-' + $1.toLowerCase();
    });
};
exports.AttestationTypeDashedNames = exports.AttestationTypeNames.map(toDashedCase);
exports.AttestationTypeIDs = Object.keys(AttestationTypeID)
    .map(function (key) { return AttestationTypeID[key]; })
    .filter(function (x) { return typeof x == 'number'; });
