// @bloomprotocol/attestations-lib/dist/src/AttestationTypes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var AttestationTypeID;
(function (AttestationTypeID) {
    AttestationTypeID[AttestationTypeID["phone"] = 0] = "phone";
    AttestationTypeID[AttestationTypeID["email"] = 1] = "email";
    AttestationTypeID[AttestationTypeID["facebook"] = 2] = "facebook";
    AttestationTypeID[AttestationTypeID["sanction-screen"] = 3] = "sanction-screen";
    AttestationTypeID[AttestationTypeID["pep-screen"] = 4] = "pep-screen";
    AttestationTypeID[AttestationTypeID["id-document"] = 5] = "id-document";
    AttestationTypeID[AttestationTypeID["google"] = 6] = "google";
    AttestationTypeID[AttestationTypeID["linkedin"] = 7] = "linkedin";
    AttestationTypeID[AttestationTypeID["twitter"] = 8] = "twitter";
    AttestationTypeID[AttestationTypeID["payroll"] = 9] = "payroll";
    AttestationTypeID[AttestationTypeID["ssn"] = 10] = "ssn";
    AttestationTypeID[AttestationTypeID["criminal"] = 11] = "criminal";
    AttestationTypeID[AttestationTypeID["offense"] = 12] = "offense";
    AttestationTypeID[AttestationTypeID["driving"] = 13] = "driving";
    AttestationTypeID[AttestationTypeID["employment"] = 14] = "employment";
    AttestationTypeID[AttestationTypeID["education"] = 15] = "education";
    AttestationTypeID[AttestationTypeID["drug"] = 16] = "drug";
    AttestationTypeID[AttestationTypeID["bank"] = 17] = "bank";
    AttestationTypeID[AttestationTypeID["utility"] = 18] = "utility";
    AttestationTypeID[AttestationTypeID["income"] = 19] = "income";
    AttestationTypeID[AttestationTypeID["assets"] = 20] = "assets";
    AttestationTypeID[AttestationTypeID["full-name"] = 21] = "full-name";
    AttestationTypeID[AttestationTypeID["birth-date"] = 22] = "birth-date";
    AttestationTypeID[AttestationTypeID["gender"] = 23] = "gender";
    AttestationTypeID[AttestationTypeID["group"] = 24] = "group";
    AttestationTypeID[AttestationTypeID["meta"] = 25] = "meta";
    AttestationTypeID[AttestationTypeID["office"] = 26] = "office";
    AttestationTypeID[AttestationTypeID["credential"] = 27] = "credential";
    AttestationTypeID[AttestationTypeID["medical"] = 28] = "medical";
    AttestationTypeID[AttestationTypeID["biometric"] = 29] = "biometric";
    AttestationTypeID[AttestationTypeID["supplemental"] = 30] = "supplemental";
    AttestationTypeID[AttestationTypeID["vouch"] = 31] = "vouch";
    AttestationTypeID[AttestationTypeID["audit"] = 32] = "audit";
    AttestationTypeID[AttestationTypeID["address"] = 33] = "address";
})(AttestationTypeID = exports.AttestationTypeID || (exports.AttestationTypeID = {}));
exports.AttestationTypes = {
    phone: {
        id: AttestationTypeID.phone,
        scoreWeight: 5,
        nameFriendly: 'Phone',
    },
    email: {
        id: 1,
        scoreWeight: 5,
        nameFriendly: 'Email',
    },
    facebook: {
        id: AttestationTypeID.facebook,
        scoreWeight: 5,
        nameFriendly: 'Facebook',
    },
    'sanction-screen': {
        id: AttestationTypeID['sanction-screen'],
        scoreWeight: 10,
        nameFriendly: 'Sanction Screen',
    },
    'pep-screen': {
        id: AttestationTypeID['pep-screen'],
        scoreWeight: 10,
        nameFriendly: 'PEP Screen',
    },
    'id-document': {
        id: AttestationTypeID['id-document'],
        scoreWeight: 20,
        nameFriendly: 'ID Document',
    },
    google: {
        id: AttestationTypeID.google,
        scoreWeight: 5,
        nameFriendly: 'Google',
    },
    linkedin: {
        id: AttestationTypeID.linkedin,
        scoreWeight: 5,
        nameFriendly: 'LinkedIn',
    },
    twitter: {
        id: AttestationTypeID.twitter,
        scoreWeight: 5,
        nameFriendly: 'Twitter',
    },
    payroll: {
        id: AttestationTypeID.payroll,
        scoreWeight: 15,
        nameFriendly: 'Payroll',
    },
    ssn: {
        id: AttestationTypeID.ssn,
        scoreWeight: 15,
        nameFriendly: 'Social Security Number',
    },
    criminal: {
        id: AttestationTypeID.criminal,
        scoreWeight: 15,
        nameFriendly: 'Criminal Records',
    },
    offense: {
        id: AttestationTypeID.offense,
        scoreWeight: 10,
        nameFriendly: 'Offense Records',
    },
    driving: {
        id: AttestationTypeID.driving,
        scoreWeight: 15,
        nameFriendly: 'Driving Records',
    },
    employment: {
        id: AttestationTypeID.employment,
        scoreWeight: 15,
        nameFriendly: 'Employment',
    },
    education: {
        id: AttestationTypeID.education,
        scoreWeight: 10,
        nameFriendly: 'Education',
    },
    drug: {
        id: AttestationTypeID.drug,
        scoreWeight: 5,
        nameFriendly: 'Drug Screen',
    },
    bank: {
        id: AttestationTypeID.bank,
        scoreWeight: 15,
        nameFriendly: 'Bank Statement',
    },
    utility: {
        id: AttestationTypeID.utility,
        scoreWeight: 10,
        nameFriendly: 'Utility Statements',
    },
    income: {
        id: AttestationTypeID.income,
        scoreWeight: 15,
        nameFriendly: 'Income Verification',
    },
    assets: {
        id: AttestationTypeID.assets,
        scoreWeight: 15,
        nameFriendly: 'Assets Verification',
    },
    'full-name': {
        id: AttestationTypeID['full-name'],
        scoreWeight: 5,
        nameFriendly: 'Full Name',
    },
    'birth-date': {
        id: AttestationTypeID['birth-date'],
        scoreWeight: 5,
        nameFriendly: 'Date of Birth',
    },
    gender: {
        id: AttestationTypeID['gender'],
        scoreWeight: 5,
        nameFriendly: 'Gender',
    },
    group: {
        id: AttestationTypeID['group'],
        scoreWeight: 5,
        nameFriendly: 'Group',
    },
    meta: {
        id: AttestationTypeID['meta'],
        scoreWeight: 20,
        nameFriendly: 'Meta-attestation',
    },
    office: {
        id: AttestationTypeID['office'],
        scoreWeight: 5,
        nameFriendly: 'Office/Position',
    },
    credential: {
        id: AttestationTypeID['credential'],
        scoreWeight: 5,
        nameFriendly: 'Credential',
    },
    medical: {
        id: AttestationTypeID['medical'],
        scoreWeight: 0,
        nameFriendly: 'Medical Information',
    },
    biometric: {
        id: AttestationTypeID['biometric'],
        scoreWeight: 20,
        nameFriendly: 'Biometric Information',
    },
    supplemental: {
        id: AttestationTypeID['supplemental'],
        scoreWeight: 0,
        nameFriendly: 'Supplemental Information',
    },
    vouch: {
        id: AttestationTypeID['vouch'],
        scoreWeight: 0,
        nameFriendly: 'Vouching',
    },
    audit: {
        id: AttestationTypeID['audit'],
        scoreWeight: 0,
        nameFriendly: 'Audit',
    },
    address: {
        id: AttestationTypeID['address'],
        scoreWeight: 5,
        nameFriendly: 'Address',
    },
};
exports.AttestationTypesByID = Object.keys(exports.AttestationTypes).reduce(function (acc, key) {
    var _a;
    var def = Object.assign({}, exports.AttestationTypes[key], {
        name: key,
    });
    return Object.assign(acc, (_a = {}, _a[def.id.toString()] = def, _a));
}, {});
exports.AttestationTypeNames = Object.keys(exports.AttestationTypes);
exports.AttestationTypeIDs = Object.keys(exports.AttestationTypes).map(function (k) { return exports.AttestationTypes[k].id; });
// Type accessors
exports.getAttestationTypeAttrib = function (typeId, attrib) {
    var val = exports.AttestationTypesByID[typeId.toString()][attrib];
    if (val === undefined || val === null) {
        throw new Error("AttestationTypeID " + typeId + " not supported");
    }
    return val;
};
var attestationTypeAccessor = function (attrib) { return function (typeId) { return exports.getAttestationTypeAttrib(typeId, attrib); }; };
exports.getAttestationTypeStr = attestationTypeAccessor('name');
exports.getBloomIDStrength = attestationTypeAccessor('scoreWeight');
exports.getFormattedName = attestationTypeAccessor('nameFriendly');
var AttestationStatus;
(function (AttestationStatus) {
    AttestationStatus["initial"] = "initial";
    AttestationStatus["ready"] = "ready";
    AttestationStatus["complete"] = "complete";
    AttestationStatus["rejected"] = "rejected";
})(AttestationStatus = exports.AttestationStatus || (exports.AttestationStatus = {}));
