// web3x/account/sign-transaction.js
"use strict";
/*
  This file is part of web3x.

  web3x is free software: you can redistribute it and/or modify
  it under the terms of the GNU Lesser General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  web3x is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public License
  along with web3x.  If not, see <http://www.gnu.org/licenses/>.
*/
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const rlp_1 = require("rlp");
const address_1 = require("../address");
const account_1 = tslib_1.__importDefault(require("../eth-lib/account"));
const bytes_1 = tslib_1.__importDefault(require("../eth-lib/bytes"));
const nat_1 = tslib_1.__importDefault(require("../eth-lib/nat"));
const rlp_2 = tslib_1.__importDefault(require("../eth-lib/rlp"));
const formatters_1 = require("../formatters");
const utils_1 = require("../utils");
async function signTransaction(tx, privateKey, eth) {
    if (!tx.gas) {
        throw new Error('gas is missing or 0');
    }
    // Resolve immediately if nonce, chainId and price are provided
    if (tx.nonce !== undefined && tx.chainId !== undefined && tx.gasPrice !== undefined) {
        return sign(tx, privateKey);
    }
    // Otherwise, get the missing info from the Ethereum Node
    const promises = [
        isNot(tx.chainId) ? eth.getId() : Promise.resolve(tx.chainId),
        isNot(tx.gasPrice) ? eth.getGasPrice() : Promise.resolve(tx.gasPrice),
        isNot(tx.nonce)
            ? eth.getTransactionCount(address_1.Address.fromString(account_1.default.fromPrivate(privateKey).address))
            : Promise.resolve(tx.nonce),
    ];
    const [chainId, gasPrice, nonce] = await Promise.all(promises);
    if (isNot(chainId) || isNot(gasPrice) || isNot(nonce)) {
        throw new Error('One of the values chainId, gasPrice, or nonce could not be fetched');
    }
    return sign({ ...tx, chainId, gasPrice, nonce }, privateKey);
}
exports.signTransaction = signTransaction;
function recoverTransaction(rawTx) {
    const values = rlp_2.default.decode(rawTx);
    const signature = account_1.default.encodeSignature(values.slice(6, 9));
    const recovery = bytes_1.default.toNumber(values[6]);
    const extraData = recovery < 35 ? [] : [bytes_1.default.fromNumber((recovery - 35) >> 1), '0x', '0x'];
    const signingData = values.slice(0, 6).concat(extraData);
    const signingDataHex = rlp_2.default.encode(signingData);
    return account_1.default.recover(utils_1.sha3(signingDataHex), signature);
}
exports.recoverTransaction = recoverTransaction;
function sign(tx, privateKey) {
    if (tx.nonce < 0 || tx.gas < 0 || tx.gasPrice < 0 || tx.chainId < 0) {
        throw new Error('gas, gasPrice, nonce or chainId is lower than 0');
    }
    const chainId = utils_1.numberToHex(tx.chainId);
    const toEncode = [
        bytes_1.default.fromNat(utils_1.numberToHex(tx.nonce)),
        bytes_1.default.fromNat(utils_1.numberToHex(tx.gasPrice)),
        bytes_1.default.fromNat(utils_1.numberToHex(tx.gas)),
        tx.to ? formatters_1.inputAddressFormatter(tx.to) : '0x',
        bytes_1.default.fromNat(tx.value ? utils_1.numberToHex(tx.value) : '0x'),
        tx.data ? utils_1.bufferToHex(tx.data) : '0x',
        bytes_1.default.fromNat(chainId || '0x1'),
        '0x',
        '0x',
    ];
    const rlpEncoded = rlp_1.encode(toEncode);
    const messageHash = utils_1.sha3(rlpEncoded);
    const signature = account_1.default.makeSigner(nat_1.default.toNumber(chainId || '0x1') * 2 + 35)(messageHash, privateKey);
    const rawTx = [...rlp_1.decode(rlpEncoded).slice(0, 6), ...account_1.default.decodeSignature(signature)];
    rawTx[6] = utils_1.makeHexEven(utils_1.trimHexLeadingZero(rawTx[6]));
    rawTx[7] = utils_1.makeHexEven(utils_1.trimHexLeadingZero(rawTx[7]));
    rawTx[8] = utils_1.makeHexEven(utils_1.trimHexLeadingZero(rawTx[8]));
    const rawTransaction = rlp_1.encode(rawTx);
    const values = rlp_1.decode(rawTransaction);
    return {
        messageHash,
        v: utils_1.trimHexLeadingZero(utils_1.bufferToHex(values[6])),
        r: utils_1.trimHexLeadingZero(utils_1.bufferToHex(values[7])),
        s: utils_1.trimHexLeadingZero(utils_1.bufferToHex(values[8])),
        rawTransaction: utils_1.bufferToHex(rawTransaction),
    };
}
exports.sign = sign;
function isNot(value) {
    return value === undefined || value === null;
}
