// web3x/providers/legacy-provider-adapter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const events_1 = require("events");
const jsonrpc_1 = require("./jsonrpc");
class LegacyProviderAdapter {
    constructor(provider) {
        this.provider = provider;
        this.eventEmitter = new events_1.EventEmitter();
    }
    subscribeToLegacyProvider() {
        if (!this.provider.on) {
            throw new Error('Legacy provider does not support subscriptions.');
        }
        this.provider.on('data', (result, deprecatedResult) => {
            result = result || deprecatedResult;
            if (!result.method) {
                return;
            }
            this.eventEmitter.emit('notification', result.params);
        });
    }
    send(method, params) {
        return new Promise((resolve, reject) => {
            const payload = jsonrpc_1.createJsonRpcPayload(method, params);
            this.provider.send(payload, (err, message) => {
                if (err) {
                    return reject(err);
                }
                if (!message) {
                    return reject(new Error('No response.'));
                }
                if (!jsonrpc_1.isValidJsonRpcResponse(message)) {
                    const msg = message.error && message.error.message
                        ? message.error.message
                        : 'Invalid JSON RPC response: ' + JSON.stringify(message);
                    return reject(new Error(msg));
                }
                const response = message;
                if (response.error) {
                    const message = response.error.message ? response.error.message : JSON.stringify(response);
                    return reject(new Error('Returned error: ' + message));
                }
                if (response.id && payload.id !== response.id) {
                    return reject(new Error(`Wrong response id ${payload.id} != ${response.id} in ${JSON.stringify(payload)}`));
                }
                resolve(response.result);
            });
        });
    }
    on(notification, listener) {
        if (notification !== 'notification') {
            throw new Error('Legacy providers only support notification event.');
        }
        if (this.eventEmitter.listenerCount('notification') === 0) {
            this.subscribeToLegacyProvider();
        }
        this.eventEmitter.on('notification', listener);
        return this;
    }
    removeListener(notification, listener) {
        if (!this.provider.removeListener) {
            throw new Error('Legacy provider does not support subscriptions.');
        }
        if (notification !== 'notification') {
            throw new Error('Legacy providers only support notification event.');
        }
        this.eventEmitter.removeListener('notification', listener);
        if (this.eventEmitter.listenerCount('notification') === 0) {
            this.provider.removeAllListeners('data');
        }
        return this;
    }
    removeAllListeners(notification) {
        this.eventEmitter.removeAllListeners('notification');
        if (this.provider.removeAllListeners) {
            this.provider.removeAllListeners('data');
        }
    }
}
exports.LegacyProviderAdapter = LegacyProviderAdapter;
