// web3x/providers/ws.js
"use strict";
/*
  This file is part of web3x.

  web3x is free software: you can redistribute it and/or modify
  it under the terms of the GNU Lesser General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  web3x is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public License
  along with web3x.  If not, see <http://www.gnu.org/licenses/>.
*/
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const isomorphic_ws_1 = tslib_1.__importDefault(require("isomorphic-ws"));
const util_1 = require("util");
const legacy_provider_adapter_1 = require("./legacy-provider-adapter");
class WebsocketProvider extends legacy_provider_adapter_1.LegacyProviderAdapter {
    constructor(url, options = {}) {
        const legacyProvider = new LegacyWebsocketProvider(url, options);
        super(legacyProvider);
        this.legacyProvider = legacyProvider;
    }
    disconnect() {
        this.legacyProvider.disconnect();
    }
}
exports.WebsocketProvider = WebsocketProvider;
class LegacyWebsocketProvider {
    constructor(url, options = {}) {
        this.responseCallbacks = {};
        this.notificationCallbacks = [];
        this.lastChunk = '';
        this.connected = false;
        this.onOpen = () => {
            this.connected = true;
        };
        this.onError = () => {
            this.timeout();
        };
        this.onClose = () => {
            this.timeout();
            this.reset();
            this.connected = false;
        };
        this.onMessage = (e) => {
            const data = typeof e.data === 'string' ? e.data : '';
            this.parseResponse(data).forEach(result => {
                let id = null;
                // get the id which matches the returned id
                if (util_1.isArray(result)) {
                    result.forEach((load) => {
                        if (this.responseCallbacks[load.id]) {
                            id = load.id;
                        }
                    });
                }
                else {
                    id = result.id;
                }
                // notification
                if (!id && result && result.method && result.method.indexOf('_subscription') !== -1) {
                    this.notificationCallbacks.forEach(callback => {
                        callback(result);
                    });
                    // fire the callback
                }
                else if (id && this.responseCallbacks[id]) {
                    this.responseCallbacks[id].callback(undefined, result);
                    delete this.responseCallbacks[id];
                }
            });
        };
        this.options = options;
        this.connection = new isomorphic_ws_1.default(url, options.protocol, options.clientOptions);
        this.connection.onopen = this.onOpen;
        this.connection.onerror = this.onError;
        this.connection.onclose = this.onClose;
        this.connection.onmessage = this.onMessage;
    }
    parseResponse(data) {
        const returnValues = [];
        // DE-CHUNKER
        const dechunkedData = data
            .replace(/\}[\n\r]?\{/g, '}|--|{') // }{
            .replace(/\}\][\n\r]?\[\{/g, '}]|--|[{') // }][{
            .replace(/\}[\n\r]?\[\{/g, '}|--|[{') // }[{
            .replace(/\}\][\n\r]?\{/g, '}]|--|{') // }]{
            .split('|--|');
        dechunkedData.forEach(data => {
            // prepend the last chunk
            if (this.lastChunk) {
                data = this.lastChunk + data;
            }
            let result = null;
            try {
                result = JSON.parse(data);
            }
            catch (e) {
                this.lastChunk = data;
                // start timeout to cancel all requests
                clearTimeout(this.lastChunkTimeout);
                this.lastChunkTimeout = setTimeout(() => {
                    this.timeout();
                    throw new Error(`Invalid response data: ${data}`);
                }, 1000 * 15);
                return;
            }
            // cancel timeout and set chunk to null
            clearTimeout(this.lastChunkTimeout);
            this.lastChunk = '';
            if (result) {
                returnValues.push(result);
            }
        });
        return returnValues;
    }
    addResponseCallback(payload, callback) {
        const id = payload.id || payload[0].id;
        const method = payload.method || payload[0].method;
        this.responseCallbacks[id] = { callback, method };
        // schedule triggering the error response if a custom timeout is set
        if (this.options.timeout) {
            setTimeout(() => {
                if (this.responseCallbacks[id]) {
                    this.responseCallbacks[id].callback(new Error('Connection timeout'), undefined);
                    delete this.responseCallbacks[id];
                }
            }, this.options.timeout);
        }
    }
    timeout() {
        for (const key in this.responseCallbacks) {
            if (this.responseCallbacks[key]) {
                this.responseCallbacks[key].callback(new Error('Connection error'), undefined);
                delete this.responseCallbacks[key];
            }
        }
    }
    send(payload, callback) {
        if (this.connection.readyState === this.connection.CONNECTING) {
            setTimeout(() => {
                this.send(payload, callback);
            }, 10);
            return;
        }
        if (this.connection.readyState !== this.connection.OPEN) {
            // tslint:disable-next-line:no-console
            console.error('connection not open on send()');
            this.onError();
            callback(new Error('connection not open'), undefined);
            return;
        }
        this.connection.send(JSON.stringify(payload));
        this.addResponseCallback(payload, callback);
    }
    on(type, callback) {
        switch (type) {
            case 'data':
                this.notificationCallbacks.push(callback);
                break;
            default:
                throw new Error('Only supports data.');
        }
    }
    removeListener(type, callback) {
        switch (type) {
            case 'data':
                const i = this.notificationCallbacks.indexOf(callback);
                if (i !== -1) {
                    this.notificationCallbacks.splice(i, 1);
                }
                break;
            default:
                throw new Error('Only supports data.');
        }
    }
    removeAllListeners(type) {
        switch (type) {
            case 'data':
                this.notificationCallbacks = [];
                break;
            default:
                throw new Error('Only supports data.');
        }
    }
    reset() {
        this.timeout();
        this.notificationCallbacks = [];
    }
    disconnect() {
        if (this.connection) {
            this.connection.close();
        }
    }
}
