// @bloomprotocol/share-kit-react/dist/src/RequestElement.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) if (e.indexOf(p[i]) < 0)
            t[p[i]] = s[p[i]];
    return t;
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
var React = __importStar(require("react"));
var share_kit_1 = require("@bloomprotocol/share-kit");
var RequestElement = /** @class */ (function (_super) {
    __extends(RequestElement, _super);
    function RequestElement(props) {
        var _this = _super.call(this, props) || this;
        _this.containerRef = React.createRef();
        return _this;
    }
    RequestElement.prototype.componentDidMount = function () {
        if (!this.containerRef.current)
            return;
        var _a = this.props, requestData = _a.requestData, shouldRenderButton = _a.shouldRenderButton, qrOptions = _a.qrOptions, buttonCallbackUrl = _a.buttonCallbackUrl;
        this.requestElementResult = share_kit_1.renderRequestElement({
            container: this.containerRef.current,
            requestData: requestData,
            qrOptions: qrOptions,
            shouldRenderButton: shouldRenderButton,
            buttonCallbackUrl: buttonCallbackUrl,
        });
    };
    RequestElement.prototype.componentDidUpdate = function (prevProps) {
        var prevRequestData = prevProps.requestData, prevQROptions = prevProps.qrOptions, prevButtonCallbackUrl = prevProps.buttonCallbackUrl;
        var _a = this.props, requestData = _a.requestData, qrOptions = _a.qrOptions, buttonCallbackUrl = _a.buttonCallbackUrl;
        if (prevRequestData !== requestData || prevQROptions !== qrOptions || prevButtonCallbackUrl !== buttonCallbackUrl) {
            this.requestElementResult.update({ requestData: requestData, qrOptions: qrOptions, buttonCallbackUrl: buttonCallbackUrl });
        }
    };
    RequestElement.prototype.componentWillUnmount = function () {
        this.requestElementResult.remove();
    };
    RequestElement.prototype.render = function () {
        var _a = this.props, requestData = _a.requestData, shouldRenderButton = _a.shouldRenderButton, qrOptions = _a.qrOptions, buttonCallbackUrl = _a.buttonCallbackUrl, rest = __rest(_a, ["requestData", "shouldRenderButton", "qrOptions", "buttonCallbackUrl"]);
        return React.createElement("div", __assign({}, rest, { ref: this.containerRef }));
    };
    return RequestElement;
}(React.Component));
exports.RequestElement = RequestElement;
