// web3x/evm/trie/trie.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const merkle_patricia_tree_1 = tslib_1.__importDefault(require("merkle-patricia-tree"));
class Trie {
    constructor(db, stateRoot) {
        this.db = db;
        this.trie = new merkle_patricia_tree_1.default(db, stateRoot);
    }
    get isCheckpoint() {
        return this.trie.isCheckpoint;
    }
    get root() {
        return this.trie.root;
    }
    async get(key) {
        return new Promise((resolve, reject) => {
            this.trie.get(key, (err, value) => {
                if (err) {
                    return reject(err);
                }
                else {
                    return resolve(value);
                }
            });
        });
    }
    async put(key, value) {
        return new Promise((resolve, reject) => this.trie.put(key, value, err => {
            if (err) {
                return reject(err);
            }
            else {
                return resolve();
            }
        }));
    }
    checkpoint() {
        this.trie.checkpoint();
    }
    commit() {
        return new Promise(resolve => this.trie.commit(resolve));
    }
    revert() {
        return new Promise(resolve => this.trie.revert(resolve));
    }
}
exports.Trie = Trie;
