// web3x/utils/hex.js
"use strict";
/*
  This file is part of web3x.

  web3x is free software: you can redistribute it and/or modify
  it under the terms of the GNU Lesser General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  web3x is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public License
  along with web3x.  If not, see <http://www.gnu.org/licenses/>.
*/
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const randombytes_1 = tslib_1.__importDefault(require("randombytes"));
const util_1 = require("util");
const address_1 = require("../address");
const bn_1 = require("./bn");
const hex_number_1 = require("./hex-number");
const hex_utf8_1 = require("./hex-utf8");
/**
 * Check if string is HEX, requires a 0x in front
 */
function isHexStrict(hex) {
    return /^(-)?0x[0-9a-f]*$/i.test(hex);
}
exports.isHexStrict = isHexStrict;
/**
 * Check if string is HEX
 */
function isHex(hex) {
    return /^(-0x|0x)?[0-9a-f]*$/i.test(hex);
}
exports.isHex = isHex;
/**
 * Auto converts any given value into it's hex representation.
 */
function toHex(value, returnType) {
    /*jshint maxcomplexity: false */
    if (util_1.isString(value) && address_1.Address.isAddress(value)) {
        return returnType ? 'address' : '0x' + value.toLowerCase().replace(/^0x/i, '');
    }
    if (util_1.isBoolean(value)) {
        return returnType ? 'bool' : value ? '0x01' : '0x00';
    }
    if (util_1.isObject(value) && !bn_1.isBN(value)) {
        return returnType ? 'string' : hex_utf8_1.utf8ToHex(JSON.stringify(value));
    }
    // if its a negative number, pass it through numberToHex
    if (util_1.isString(value)) {
        if (value.indexOf('-0x') === 0 || value.indexOf('-0X') === 0) {
            return returnType ? 'int256' : hex_number_1.numberToHex(value);
        }
        else if (value.indexOf('0x') === 0 || value.indexOf('0X') === 0) {
            return returnType ? 'bytes' : value;
        }
        else if (!isFinite(+value)) {
            return returnType ? 'string' : hex_utf8_1.utf8ToHex(value);
        }
    }
    return returnType ? (value < 0 ? 'int256' : 'uint256') : hex_number_1.numberToHex(value);
}
exports.toHex = toHex;
function randomHex(size) {
    if (size > 65536) {
        throw new Error('Requested too many random bytes.');
    }
    return '0x' + randombytes_1.default(size).toString('hex');
}
exports.randomHex = randomHex;
function randomBuffer(size) {
    if (size > 65536) {
        throw new Error('Requested too many random bytes.');
    }
    return randombytes_1.default(size);
}
exports.randomBuffer = randomBuffer;
function trimHexLeadingZero(hex) {
    return hex.replace(/^0x0*/, '0x');
}
exports.trimHexLeadingZero = trimHexLeadingZero;
function makeHexEven(hex) {
    return hex.length % 2 === 1 ? hex.replace('0x', '0x0') : hex;
}
exports.makeHexEven = makeHexEven;
