// web3x/utils/sign.js
"use strict";
/*
  This file is part of web3x.

  web3x is free software: you can redistribute it and/or modify
  it under the terms of the GNU Lesser General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  web3x is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public License
  along with web3x.  If not, see <http://www.gnu.org/licenses/>.
*/
Object.defineProperty(exports, "__esModule", { value: true });
const address_1 = require("../address");
const account_1 = require("../eth-lib/account");
const hash_message_1 = require("./hash-message");
function sign(data, privateKey) {
    const messageHash = hash_message_1.hashMessage(data);
    const signature = account_1.sign(messageHash, privateKey);
    const vrs = account_1.decodeSignature(signature);
    return {
        message: data,
        messageHash,
        v: vrs[0],
        r: vrs[1],
        s: vrs[2],
        signature,
    };
}
exports.sign = sign;
function recoverFromSignature(signature) {
    const { messageHash, v, r, s } = signature;
    return recoverFromSigString(messageHash, account_1.encodeSignature([v, r, s]), true);
}
exports.recoverFromSignature = recoverFromSignature;
function recoverFromVRS(message, v, r, s, prefixed = false) {
    if (!prefixed) {
        message = hash_message_1.hashMessage(message);
    }
    return recoverFromSigString(message, account_1.encodeSignature([v, r, s]), true);
}
exports.recoverFromVRS = recoverFromVRS;
function recoverFromSigString(message, signature, preFixed = false) {
    if (!preFixed) {
        message = hash_message_1.hashMessage(message);
    }
    return address_1.Address.fromString(account_1.recover(message, signature));
}
exports.recoverFromSigString = recoverFromSigString;
function recover(...args) {
    switch (args.length) {
        case 1:
            return recoverFromSignature(args[0]);
        case 2:
        case 3:
            return recoverFromSigString(args[0], args[1], args[2]);
        case 4:
        case 5:
            return recoverFromVRS(args[0], args[1], args[2], args[3], args[4]);
    }
    throw new Error('Cannot determine recovery function');
}
exports.recover = recover;
