// web3x/formatters/log-response-formatter.js
"use strict";
/*
  This file is part of web3x.

  web3x is free software: you can redistribute it and/or modify
  it under the terms of the GNU Lesser General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  web3x is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public License
  along with web3x.  If not, see <http://www.gnu.org/licenses/>.
*/
Object.defineProperty(exports, "__esModule", { value: true });
const util_1 = require("util");
const address_1 = require("../address");
const utils_1 = require("../utils");
function fromRawLogResponse(log) {
    let id = log.id || null;
    // Generate a custom log id.
    if (typeof log.blockHash === 'string' &&
        typeof log.transactionHash === 'string' &&
        typeof log.logIndex === 'string') {
        const shaId = utils_1.sha3(log.blockHash.replace('0x', '') + log.transactionHash.replace('0x', '') + log.logIndex.replace('0x', ''));
        id = 'log_' + shaId.replace('0x', '').substr(0, 8);
    }
    const blockNumber = log.blockNumber !== null ? utils_1.hexToNumber(log.blockNumber) : null;
    const transactionIndex = log.transactionIndex !== null ? utils_1.hexToNumber(log.transactionIndex) : null;
    const logIndex = log.logIndex !== null ? utils_1.hexToNumber(log.logIndex) : null;
    const address = util_1.isString(log.address) ? address_1.Address.fromString(log.address) : log.address;
    return { log, id, blockNumber, transactionIndex, logIndex, address };
}
exports.fromRawLogResponse = fromRawLogResponse;
function toRawLogResponse(log) {
    const { id, blockNumber, transactionIndex, logIndex, address } = log;
    return {
        // ...log,
        id: id ? id : undefined,
        blockNumber: util_1.isNumber(blockNumber) ? utils_1.numberToHex(blockNumber) : null,
        transactionIndex: util_1.isNumber(transactionIndex) ? utils_1.numberToHex(transactionIndex) : null,
        logIndex: util_1.isNumber(logIndex) ? utils_1.numberToHex(logIndex) : null,
        address: address.toString().toLowerCase(),
    };
}
exports.toRawLogResponse = toRawLogResponse;
