// web3x/formatters/transaction-response-formatter.js
"use strict";
/*
  This file is part of web3x.

  web3x is free software: you can redistribute it and/or modify
  it under the terms of the GNU Lesser General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  web3x is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public License
  along with web3x.  If not, see <http://www.gnu.org/licenses/>.
*/
Object.defineProperty(exports, "__esModule", { value: true });
const address_1 = require("../address");
const utils_1 = require("../utils");
const output_big_number_formatter_1 = require("./output-big-number-formatter");
function fromRawTransactionResponse(tx) {
    return {
        tx,
        blockNumber: tx.blockNumber ? utils_1.hexToNumber(tx.blockNumber) : null,
        transactionIndex: tx.transactionIndex ? utils_1.hexToNumber(tx.transactionIndex) : null,
        nonce: utils_1.hexToNumber(tx.nonce),
        gas: utils_1.hexToNumber(tx.gas),
        gasPrice: output_big_number_formatter_1.outputBigNumberFormatter(tx.gasPrice),
        value: output_big_number_formatter_1.outputBigNumberFormatter(tx.value),
        to: tx.to ? address_1.Address.fromString(tx.to) : null,
        from: address_1.Address.fromString(tx.from),
    };
}
exports.fromRawTransactionResponse = fromRawTransactionResponse;
function toRawTransactionResponse(tx) {
    return {
        // ...tx,
        blockNumber: tx.blockNumber ? utils_1.numberToHex(tx.blockNumber) : null,
        transactionIndex: tx.transactionIndex ? utils_1.numberToHex(tx.transactionIndex) : null,
        nonce: utils_1.numberToHex(tx.nonce),
        gas: utils_1.numberToHex(tx.gas),
        gasPrice: utils_1.numberToHex(tx.gasPrice),
        value: utils_1.numberToHex(tx.value),
        to: tx.to ? tx.to.toString() : null,
        from: tx.from.toString(),
    };
}
exports.toRawTransactionResponse = toRawTransactionResponse;
