// web3x/formatters/transaction-receipt-formatter.js
"use strict";
/*
  This file is part of web3x.

  web3x is free software: you can redistribute it and/or modify
  it under the terms of the GNU Lesser General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  web3x is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public License
  along with web3x.  If not, see <http://www.gnu.org/licenses/>.
*/
Object.defineProperty(exports, "__esModule", { value: true });
const util_1 = require("util");
const address_1 = require("../address");
const utils_1 = require("../utils");
const log_response_formatter_1 = require("./log-response-formatter");
function fromRawTransactionReceipt(receipt) {
    if (!receipt || !receipt.blockHash) {
        return null;
    }
    return {
        receipt,
        to: receipt.to ? address_1.Address.fromString(receipt.to) : undefined,
        from: address_1.Address.fromString(receipt.from),
        blockNumber: utils_1.hexToNumber(receipt.blockNumber),
        transactionIndex: utils_1.hexToNumber(receipt.transactionIndex),
        cumulativeGasUsed: utils_1.hexToNumber(receipt.cumulativeGasUsed),
        gasUsed: utils_1.hexToNumber(receipt.gasUsed),
        logs: util_1.isArray(receipt.logs) ? receipt.logs.map(log_response_formatter_1.fromRawLogResponse) : undefined,
        contractAddress: receipt.contractAddress ? address_1.Address.fromString(receipt.contractAddress) : undefined,
        status: util_1.isString(receipt.status) ? Boolean(utils_1.hexToNumber(receipt.status)) : undefined,
    };
}
exports.fromRawTransactionReceipt = fromRawTransactionReceipt;
function toRawTransactionReceipt(receipt) {
    return {
        receipt,
        to: receipt.to ? receipt.to.toString() : undefined,
        from: receipt.from.toString(),
        blockNumber: utils_1.numberToHex(receipt.blockNumber),
        transactionIndex: utils_1.numberToHex(receipt.transactionIndex),
        cumulativeGasUsed: utils_1.numberToHex(receipt.cumulativeGasUsed),
        gasUsed: utils_1.numberToHex(receipt.gasUsed),
        logs: util_1.isArray(receipt.logs) ? receipt.logs.map(log_response_formatter_1.toRawLogResponse) : undefined,
        contractAddress: receipt.contractAddress ? receipt.contractAddress.toString() : undefined,
        status: receipt.status !== undefined ? utils_1.numberToHex(receipt.status ? 1 : 0) : undefined,
    };
}
exports.toRawTransactionReceipt = toRawTransactionReceipt;
