// web3x/evm/provider/evm-provider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const bigint_buffer_1 = require("bigint-buffer");
const events_1 = require("events");
const levelup_1 = tslib_1.__importDefault(require("levelup"));
const address_1 = require("../../address");
const formatters_1 = require("../../formatters");
const utils_1 = require("../../utils");
const blockchain_1 = require("../blockchain");
const vm_1 = require("../vm");
const world_1 = require("../world");
const handle_call_1 = require("./handle-call");
const handle_get_logs_1 = require("./handle-get-logs");
const handle_get_transaction_receipt_1 = require("./handle-get-transaction-receipt");
const handle_send_transaction_1 = require("./handle-send-transaction");
class EvmProvider extends events_1.EventEmitter {
    constructor(worldState, blockchain, options = {}) {
        super();
        this.worldState = worldState;
        this.blockchain = blockchain;
        this.options = options;
        this.subscriptions = {};
        this.nextSubscriptionId = 0;
    }
    static fromEvmProvider(provider, options) {
        return new EvmProvider(provider.worldState, provider.blockchain, options);
    }
    static async fromDb(db, options) {
        const worldState = await world_1.WorldState.fromDb(db);
        const blockchain = await blockchain_1.Blockchain.fromDb(db);
        return new EvmProvider(worldState, blockchain, options);
    }
    static async fromLocalDb(name, options) {
        const leveljs = require('level-js');
        return await EvmProvider.fromDb(levelup_1.default(leveljs(name)), options);
    }
    static async eraseLocalDb(name) {
        return new Promise((resolve, reject) => {
            const req = indexedDB.deleteDatabase(`level-js-${name}`);
            req.onsuccess = resolve;
            req.onerror = reject;
            req.onblocked = reject;
        });
    }
    async send(method, params) {
        // console.log(method);
        // console.log(params);
        switch (method) {
            case 'eth_gasPrice':
                return utils_1.numberToHex(50000);
        }
        if (!params || !params[0]) {
            throw new Error();
        }
        switch (method) {
            case 'eth_sendTransaction':
                if (!this.wallet) {
                    throw new Error('No wallet available for signing transactions.');
                }
                return await handle_send_transaction_1.handleSendTransaction(this.worldState, this.blockchain, formatters_1.fromRawTransactionRequest(params[0]), this.wallet, this.options.blockDelay);
            case 'eth_call':
                return utils_1.bufferToHex(await handle_call_1.handleCall(this.worldState, formatters_1.fromRawCallRequest(params[0])));
            case 'eth_getTransactionReceipt':
                return await handle_get_transaction_receipt_1.handleGetTransactionReceipt(this.blockchain, params[0]);
            case 'eth_getCode':
                return utils_1.bufferToHex(await vm_1.getAccountCode(this.worldState, address_1.Address.fromString(params[0])));
            case 'eth_getLogs':
                return (await handle_get_logs_1.getLogs(this.blockchain, formatters_1.fromRawLogRequest(params[0]))).map(formatters_1.toRawLogResponse);
            case 'eth_subscribe':
                return utils_1.numberToHex(this.subscribe(params[0], params[1]));
            case 'eth_unsubscribe':
                return this.unsubscribe(params[0]);
        }
    }
    subscribe(event, params) {
        const id = utils_1.numberToHex(this.nextSubscriptionId++);
        if (event === 'logs') {
            const listener = logs => this.handleLogs(id, logs);
            this.blockchain.on('logs', listener);
            this.subscriptions[id] = { event, params, listener };
        }
        if (event === 'newHeads') {
            const listener = (block, hash) => this.handleNewHead(id, block, hash);
            this.blockchain.on('newHeads', listener);
            this.subscriptions[id] = { event, params, listener };
        }
        return id;
    }
    unsubscribe(id) {
        const sub = this.subscriptions[id];
        if (!sub) {
            return false;
        }
        delete this.subscriptions[id];
        this.blockchain.removeListener(sub.event, sub.listener);
        return true;
    }
    handleLogs(subscription, logResponse) {
        this.emit('notification', { subscription, result: formatters_1.toRawLogResponse(logResponse) });
    }
    handleNewHead(subscription, block, hash) {
        const blockResponse = {
            hash,
            ...block,
            difficulty: block.difficulty.toString(),
            gasLimit: Number(block.gasLimit),
            gasUsed: Number(block.gasUsed),
            nonce: bigint_buffer_1.toBufferBE(BigInt(block.nonce), 8),
        };
        this.emit('notification', { subscription, result: formatters_1.toRawBlockHeaderResponse(blockResponse) });
    }
    on(notification, listener) {
        return super.on(notification, listener);
    }
    removeListener(notification, listener) {
        return super.removeListener(notification, listener);
    }
    removeAllListeners(notification) {
        return super.removeAllListeners(notification);
    }
}
exports.EvmProvider = EvmProvider;
