// web3x/evm/provider/handle-get-transaction-receipt.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const bigint_buffer_1 = require("bigint-buffer");
const rlp = tslib_1.__importStar(require("rlp"));
const address_1 = require("../../address");
const formatters_1 = require("../../formatters");
const utils_1 = require("../../utils");
const tx_1 = require("../tx");
async function handleGetTransactionReceipt(blockchain, transactionHash) {
    const txHash = utils_1.hexToBuffer(transactionHash);
    const receipt = await blockchain.getTransactionReceipt(txHash);
    if (!receipt) {
        return null;
    }
    const tx = await blockchain.getMinedTransaction(txHash);
    const { to, nonce } = tx;
    // TODO: Store from in tx so no need to recover? This is slow.
    const from = tx_1.recoverTransaction(tx);
    const { cumulativeGasUsed, logs, status } = receipt;
    const receiptLogs = logs.map((log, logIndex) => ({
        id: null,
        removed: false,
        logIndex,
        blockNumber: 0,
        blockHash: '0',
        transactionHash,
        transactionIndex: 0,
        address: log.address,
        data: utils_1.bufferToHex(log.data),
        topics: log.topics.map(utils_1.bufferToHex),
    }));
    const txReceipt = {
        transactionHash,
        transactionIndex: 0,
        blockHash: '0',
        blockNumber: 0,
        from,
        to,
        cumulativeGasUsed: Number(cumulativeGasUsed),
        gasUsed: Number(cumulativeGasUsed),
        contractAddress: !to ? getContractAddress(from, nonce) : undefined,
        logs: receiptLogs,
        status,
    };
    return formatters_1.toRawTransactionReceipt(txReceipt);
}
exports.handleGetTransactionReceipt = handleGetTransactionReceipt;
function getContractAddress(from, nonce) {
    return new address_1.Address(utils_1.sha3Buffer(rlp.encode([from.toBuffer(), bigint_buffer_1.toBufferBE(nonce, 32)])).slice(12));
}
