// web3x/evm/provider/handle-get-logs.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const util_1 = require("util");
const utils_1 = require("../../utils");
async function getLogs(blockchain, logRequest) {
    const { toBlock, fromBlock, address, topics } = logRequest;
    const chainTip = await blockchain.getChaintip();
    const toBlockNum = toBlock !== undefined ? blockTypeToNumber(toBlock, chainTip) : chainTip.number;
    const fromBlockNum = fromBlock !== undefined ? blockTypeToNumber(fromBlock, chainTip) : chainTip.number;
    const addresses = address ? (util_1.isArray(address) ? address : [address]) : [];
    const result = await blockchain.getLogs(addresses, topicsToBuffers(topics), fromBlockNum, toBlockNum);
    return result.map(({ block, blockHash, transactionIndex, transactionHash, log: { address, data, topics } }) => ({
        id: null,
        removed: false,
        logIndex: 0,
        blockNumber: block.number,
        blockHash: utils_1.bufferToHex(blockHash),
        transactionHash: utils_1.bufferToHex(transactionHash),
        transactionIndex,
        address,
        data: utils_1.bufferToHex(data),
        topics: topics.map(utils_1.bufferToHex),
    }));
}
exports.getLogs = getLogs;
function blockTypeToNumber(block, chainTip) {
    switch (block) {
        case 'genesis':
            return 0;
        case 'latest':
            return chainTip.number;
        case 'pending':
            throw new Error('Pending block type not supported.');
        default:
            return block;
    }
}
function topicsToBuffers(topics) {
    return topics.map(topic => {
        if (topic === null) {
            return null;
        }
        topic = util_1.isArray(topic) ? topic : [topic];
        return topic.map(utils_1.hexToBuffer);
    });
}
