// web3x/evm/provider/handle-send-transaction.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const sign_transaction_1 = require("../../account/sign-transaction");
const utils_1 = require("../../utils");
const tx_1 = require("../tx");
const vm_1 = require("../vm");
async function handleSendTransaction(worldState, blockchain, txRequest, wallet, blockDelay = 0) {
    const { from, to, gas = 200000, gasPrice, value = 0, data } = txRequest;
    const nonce = txRequest.nonce ? BigInt(txRequest.nonce) : await worldState.getTransactionCount(from);
    const fromAccount = wallet.get(from);
    if (!fromAccount) {
        throw new Error(`Unknown address: ${from}`);
    }
    const signTxRequest = {
        chainId: 1,
        to,
        gas,
        gasPrice,
        value,
        data,
        nonce: nonce.toString(),
    };
    // TODO: Move sign function somewhere better (out of account module)?
    const { v, r, s } = sign_transaction_1.sign(signTxRequest, fromAccount.privateKey);
    const tx = {
        nonce,
        to,
        dataOrInit: data,
        gasPrice: BigInt(gasPrice),
        gasLimit: BigInt(gas),
        value: BigInt(value),
        v,
        r,
        s,
    };
    const txHash = utils_1.sha3(tx_1.serializeTx(tx));
    setTimeout(async () => {
        const result = await vm_1.executeTransaction(worldState, tx, from);
        const receipt = {
            cumulativeGasUsed: BigInt(gas) - result.remainingGas,
            logs: result.txSubstrate.logs,
            logsBloomFilter: Buffer.of(),
            status: result.status,
        };
        await blockchain.mineTransactions(await worldState.getStateRoot(), [tx], [receipt]);
    }, blockDelay);
    return txHash;
}
exports.handleSendTransaction = handleSendTransaction;
